[1단계 - 출석 구현] 벨로(방예혁) 미션 제출합니다.

YehyeokBang opened the pull request:
안녕하세요 엘리 😄 

이번 리뷰어로 만나게 되어 영광입니다!!
시간 내어 리뷰해주셔서 감사합니다! 👍 👍 

<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->

## 체크 리스트

- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  - https://prolog.techcourse.co.kr/studylogs/3908

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?

### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [x] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.

<!-- 이유 작성 -->

페어와 함께 둘만 알 수 있도록 하지 않고 (조금 노력하면) 다른 사람도 읽을 수 있으면 좋겠다. 라는 기준을 만들고 서로 합의하에 코드를 작성했습니다. 다만, 미션 요구 사항을 구현할 수록 책임이 모호해지는 경우가 있어서 3점을 선택했습니다.
  
## 어떤 부분에 집중하여 리뷰해야 할까요?

<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 피드백받고 싶은 내용을 공유해주세요.  
예를 들어, 가장 고민했던 점이나 여전히 어려운 부분, 그리고 이에 대한 생각을 적을 수 있습니다. -->

## View가 Enum을 알고 있어도 되는가?

처음에는 프로그램 전반에서 사용되는 규칙(예: 몇 번 결석하면 제적 대상) 같은 것을 Enum으로 정의하고,
View에서 이를 직접 사용하도록 구현했습니다.

이렇게 하면 수정이 필요할 때 Enum만 변경하면 되므로, 유지보수성이 좋다고 생각했습니다.

```java
// 출결 경고 수준 Enum
public enum AttendanceWarningLevel {

    CLEAN((totalAbsentCount) -> totalAbsentCount < 2),
    WARNING((totalAbsentCount) -> totalAbsentCount == 2), // 경고
    MEETING((totalAbsentCount) -> totalAbsentCount >= 3 && totalAbsentCount <= 5), // 면담
    EXPULSION((totalAbsentCount) -> totalAbsentCount > 5), // 제적
    ;

    private final Function<Integer, Boolean> isMatch;

    // 생성자 생략

    public static AttendanceWarningLevel judge(int lateCount, int absenceCount) {
        // isMatch 함수를 사용하여 지각 + 결석 횟수로 경고 수준을 판단한다.   
    }
}
```

하지만, Enum 스스로 시스템의 중요한 동작을 결정하는 로직을 수행한다면,
중요한 도메인 로직이고, View가 이를 직접 참조하면 유연한 변경이 어렵지 않을까? 라는 생각이 들었습니다.

그럼에도 불구하고 그대로 Enum을 View로 전달한 이유는 명확한 방법을 찾지 못했기 때문입니다.
특히, DTO로 전달하여 그대로 출력하게 하는 방법을 생각했지만,
Enum을 그대로 전달했을 때보다 이점을 체감하지 못했고,
오히려 복잡성만 늘어났다고 느꼈기 때문입니다.

반대로 Enum을 단순한 데이터 표현 용도로만 활용한다면,
그대로 전달해도 될 것 같다는 생각이 드네요.

제 고민, 생각 흐름에 대한 엘리의 의견도 궁금해요! 😄 

## 출석하지 않은 경우, 출석 객체를 만들어야 할까?

출석 기록을 출력할 때, 
출석이 가능한(주말 또는 공휴일이 아닌) 날에 출석 기록이 없는 경우,
시간을 `--:--`로 표시해야 한다는 요구사항이 있었습니다.

이때, 출석을 하지 않은 경우에도 출석 객체를 만들어야 할까? 라는 고민이 들었습니다.

###  고민했던 두 가지 방식

- 출석 객체를 미리 생성하는 방법

    프로그램 실행 시, 모든 출석 가능한 날을 확인하여 출석 기록이 없는 경우 결석 객체를 만들어 관리한다.
    하지만, 이 경우 Attendance 객체의 dateTime 필드가 null이 될 가능성이 크고,
    출석 여부를 확인하는 모든 로직에서 null을 고려해야 한다는 문제가 있습니다.
    
- 출력 시점에서 처리하는 방법 (선택한 방식)

    출석한 기록만 저장하고, 출력할 때 출석 가능한 날짜 목록과 비교하여 누락된 날짜는 `--:--`로 표시하도록 구현.
    즉, Attendance 객체를 불필요하게 생성하는 대신, 출력용 데이터를 가공하는 방식을 선택.

저희는 두 번째 방식을 선택했습니다.

출석을 하지 않은 경우, 출석 객체를 생성하는 것이 어색하다고 느꼈으며,
`Attendance` 객체가 `null`을 가질 경우, 모든 관련 로직에서 `null 처리를 고려`해야 하는 복잡성이 생긴다고 느꼈습니다.
실제로 저장하는 데이터는 출석한 기록만 있으면 충분하며, 출력 시점에서 `--:--`로 대체하는 방식이 더 직관적이라고 판단했습니다.

> 출석 시간(LocalDateTime)에 `null` 대신 `private static final ABSENCE_TIME = LocalDateTime.MIN;` 처럼
상수로 특정 시간을 넣어두고 관리해 보려고 구현해봤습니다.
이 방법은 null에 접근하는 것은 막을 수 있지만, 결국 페어끼리만 아는 하나의 컨텍스트가 되는 것 같아서 포기했습니다.

언제나 어떤 객체의 필드가 `null`이 된다는 것은 피하고 싶은 것 같아요.
상황(마감 기한, 잘못된 설계 등)마다 다르겠지만, 유사한 상황에서 `null` 필드를 피할 수 있는 방법이 있을까요?

## 출석을 수정한다. 또는 새로운 출석을 만들어 대체한다.

구현은 새로운 출석을 만들어 대체했습니다. 그러나 현재 상황(및 구조)에서 무엇이 더 좋은지 답을 내리지 못했습니다.
- 먼저 구현했던 것이 출석 객체가 불변이었고, 답을 구하지 못해 수정 기능을 구현할 때 그대로 대체하는 방식으로 구현했습니다.

지금과 같은 요구사항이라면 엘리는 어떤 방식을 선택했을 것 같나요? 엘리가 생각하는 장단점이 궁금해요! 😄 
- final 키워드를 제거 후 객체를 수정하는 방식
- 불변으로 관리하며 기존 출석 객체를 지우고 새로운(수정할) 출석 객체를 추가

## 잘 부탁드립니다!

매일 페어와 함께 코드를 작성하기 전에 대화로 짧게 회고하고 당일 지켜야할 약속을 정하고 코드를 작성했습니다.
- 예시: 3곳 이상 중복되고, 모두가 불편함을 느끼기 전에는 추상화 및 분리를 금지한다. (마감 기한 우선) 

한번 풀어봤던 문제이지만, 처음 알게된 페어와 함께 돌아가며 코드를 작성하는 것은 생각보다 많이 어렵네요.
그러나 더 많은 순간에 함께 고민할 수 있게 되어서 생각의 정리를 할 수 있는 좋은 시간이었습니다.

마감 기한, TDD, 새로운 페어 등 많은 이슈를 이겨내고 돌아가는 출석 프로그램을 만든 것이 큰 동기부여가 되네요.
어딘가 많이 불편한 코드이지만, 잘 부탁드립니다. 😄

YehyeokBang committed:
docs: 구현할 기능 목록 작성

Co-authored-by: abc5259 <dlwogns3413@naver.com>

YehyeokBang committed:
feat: 커맨드 입력 기능 구현

Co-authored-by: abc5259 <dlwogns3413@naver.com>

YehyeokBang committed:
feat: 출석 불가능 여부 검증 기능 구현

Co-authored-by: abc5259 <dlwogns3413@naver.com>

YehyeokBang committed:
feat: 출석 내역 불러오기 기능 구현

- csv 파일을 읽어옵니다.
- 중복 관련 검증을 추가합니다.

Co-authored-by: abc5259 <dlwogns3413@naver.com>

YehyeokBang committed:
feat: 출석 확인을 위한 닉네임 입력 기능 구현

Co-authored-by: abc5259 <dlwogns3413@naver.com>

YehyeokBang committed:
feat: 출석 확인을 위한 등교 시간 입력 기능 구현

Co-authored-by: abc5259 <dlwogns3413@naver.com>

YehyeokBang committed:
feat: 출석 내역 클래스 수정 기능 구현

Co-authored-by: abc5259 <dlwogns3413@naver.com>

YehyeokBang committed:
feat: 출석 수정 입력 기능 구현

Co-authored-by: abc5259 <dlwogns3413@naver.com>

YehyeokBang committed:
feat: 출결 유형 판단 기능 구현

Co-authored-by: abc5259 <dlwogns3413@naver.com>

YehyeokBang committed:
feat: 크루와 날짜로 출석 기록 조회 기능 구현

Co-authored-by: abc5259 <dlwogns3413@naver.com>

YehyeokBang committed:
feat: 출석 수정 기능 구현

Co-authored-by: abc5259 <dlwogns3413@naver.com>

YehyeokBang committed:
feat: 출석 기록 조회 기능 구현

Co-authored-by: abc5259 <dlwogns3413@naver.com>

YehyeokBang committed:
feat: 출석 기록 출력 및 각 횟수 출력 기능 구현

Co-authored-by: abc5259 <dlwogns3413@naver.com>

YehyeokBang committed:
feat: 크루별 출석 기록 확인 기능 구현

- 제적 위험 여부 출력을 추가합니다.

Co-authored-by: abc5259 <dlwogns3413@naver.com>

YehyeokBang committed:
feat: 제적 위험자 확인 기능 구현

Co-authored-by: abc5259 <dlwogns3413@naver.com>

YehyeokBang committed:
feat: 프로그램 종료 기능 구현

- 종료하지 않는다면 입력을 계속 받습니다.

Co-authored-by: abc5259 <dlwogns3413@naver.com>

YehyeokBang committed:
fix: 크루의 출석 기록 수정 안되던 오류 수정

- Set에서 remove 후 add하는 방식으로 해결합니다.

Co-authored-by: abc5259 <dlwogns3413@naver.com>

YehyeokBang committed:
refactor: 출력 기능 책임 분리

Co-authored-by: abc5259 <dlwogns3413@naver.com>

YehyeokBang committed:
refactor: 출석 유형 스스로 판단할 수 있도록 수정

Co-authored-by: abc5259 <dlwogns3413@naver.com>

YehyeokBang committed:
feat: 입력에 문제가 생긴 경우 예외 처리 구현

Co-authored-by: abc5259 <dlwogns3413@naver.com>

YehyeokBang committed:
refactor: 출석 수정 시 기존 출석이 반환되도록 수정

Co-authored-by: abc5259 <dlwogns3413@naver.com>

YehyeokBang committed:
test: 크루 동등성 테스트 작성

Co-authored-by: abc5259 <dlwogns3413@naver.com>

YehyeokBang committed:
refactor: 코드 가독성 개선

- 들여쓰기 및 라인 수 규칙을 준수합니다.

Co-authored-by: abc5259 <dlwogns3413@naver.com>

YehyeokBang committed:
fix: 출석 기록을 전날까지만 출력하도록 수정

- 입력을 위한 문구 오타를 수정합니다.

Co-authored-by: abc5259 <dlwogns3413@naver.com>

YehyeokBang committed:
feat: 시간 형식이 다른 경우 예외 처리 구현

Co-authored-by: abc5259 <dlwogns3413@naver.com>

YehyeokBang committed:
refactor: 메서드 분리

Co-authored-by: abc5259 <dlwogns3413@naver.com>

yebink commented on src/main/java/attendance/model/AttendanceType.java:
```diff
@@ -0,0 +1,33 @@
+
+public enum AttendanceType {
+
+    OK((difMinutes) -> difMinutes <= 5),
```
불필요한 괄호를 제거해볼까요~?

```suggestion
    OK(difMinutes -> difMinutes <= 5),
```

yebink replied:
`모든 원시 값과 문자열을 포장한다` 요구사항도 반영해볼까요?

해당 부분 외에도 전반적으로 체크해주세요 😄

yebink commented on src/main/java/attendance/controller/AttendanceController.java:
```diff
@@ -0,0 +1,237 @@
+        return command != Command.QUIT;
+    }
+
+    private void logic(Command command, LocalDateTime now) {
```
`logic` 이라는 이름만으로는 어떤 역할을 하는 함수인지 알기가 모호하네요 🤔 
더 적절한 네이밍을 고민해봅시다!

yebink commented on src/main/java/attendance/view/OutputView.java:
```diff
@@ -0,0 +1,106 @@
+
+public class OutputView {
+
+    public static final String ERROR_PREFIX = "[ERROR] ";
```
`OutputView` 에서만 사용하는 변수네요. private 으로 변경해볼까요?

YehyeokBang replied:
꼼꼼하게 봐주셔서 감사합니다! 

IntelliJ의 상수 추출 단축키를 사용했었는데, 
기본 접근 제어자 설정이 public이었네요. 추출 후 놓쳤던 것 같아요.

<img width="301" alt="image" src="https://github.com/user-attachments/assets/e6ed189d-d7f3-4511-a14c-a0d7ef4ab1f6" />

기본 설정을 private로 변경했고, 
다음에도 이러지 않도록 다시 한번 확인하는 습관을 가져보겠습니다! 😄

yebink commented on src/main/java/attendance/controller/AttendanceController.java:
```diff
@@ -0,0 +1,237 @@
+        return summary -> summary.crew().getNickname();
+    }
+
+    public record CrewAttendanceSummary(
+            Crew crew,
+            int lateCount,
+            int absenceCount,
+            AttendanceWarningLevel level
+    ) {
+    }
```
`CrewAttendanceSummary`를 `AttendanceController` 내부에서 선언하신 이유가 있나요?

YehyeokBang replied:
처음에는 Controller에서만 사용되는 DTO라고 생각하여 내부 클래스로 선언했었습니다.

하지만 이후 View에서도 해당 데이터를 사용하게 되면서 public으로 열게 되었고,
이 과정에서 내부 클래스에 두었던 사실을 간과했던 것 같습니다.

다시 보니, 다른 패키지에서도 접근해야 하는 클래스라면 별도로 분리하는 것이 더 적절하겠다는 생각이 들었습니다!
서로 다른 곳에서 사용하는 public 클래스가 하나의 클래스 내부에 작성되어 있으면 
찾기, 또한 그 의도를 파악하기도 어려울 것 같네요. 👍

yebink commented on src/main/java/attendance/controller/AttendanceController.java:
```diff
@@ -0,0 +1,237 @@
+        } while (isQuit);
+    }
+
+    private boolean start(LocalDateTime now) {
```
파라미터명이 now 로 고정되어 있네요.

요구사항이 "오늘 날짜를 입력하도록" 바뀌어도 해당 변수명이 유효할까요?

YehyeokBang replied:
팀(혹은 개인)의 생각에 따라 유효할 수도 아닐 수도 있을 것 같습니다.

변수명이 now일 때, 이를 `현재 시각(혹은 시각)` 으로 해석할 수도 있지만,
`start()` 메서드 입장에서 보면 단순히 주어진 날짜를 오늘로 간주 하는 개념으로도 볼 수 있을 것 같다는 생각을 했어요.

하지만 `now`라는 이름이 일반적으로 `현재 날짜`을 의미하는 경우가 훨씬 많다고 생각하기 때문에
요구사항이 바뀌었을 때 반드시 `현재 시간`을 입력해야 할 것 같은 인상을 줄 확률이 많이 높을 것 같아요.

그렇다면, `기준 날짜` 혹은 `호출할 날짜` 과 같은 의미가 더 명확할 것 같네요.
- 이렇게 변경하면 `메서드를 사용하는 입장`에서 전달해야 하는 파라미터의 의도를 쉽게 이해할 수 있을 것 같아요.

오해할 확률이 월등히 높다고 생각하기 때문에 변경하면 좋을 것 같네요! 감사합니다 👍

yebink commented:
안녕하세요 벨로!
출석 미션 리뷰를 맡은 엘리입니다.

주신 질문에 대한 답변 드려요.

1. View에서 Enum 직접 참조 관련
* 우선 Enum 에 로직을 아주 잘 녹여내셨네요! 💯 
  * 가독성도 좋고, 로직이 한눈에 들어와서 코드를 이해하기 편했습니다. 👍 
* DTO 를 사용했을 때 이점이 없다는 부분에 동의합니다. 
* Enum도 엄연히 객체이니, 질문을 해보는 건 어떨까요?
```java

```

yebink commented on src/main/java/attendance/view/OutputView.java:
```diff
@@ -0,0 +1,106 @@
+        return "결석";
+    }
+
+    private String displayAttendanceWarningLevel(AttendanceWarningLevel level) {
+        if (level == AttendanceWarningLevel.WARNING) {
+            return "경고";
+        }
+        if (level == AttendanceWarningLevel.MEETING) {
+            return "면담";
+        }
+        return "제적";
+    }
```
`AttendanceWarningLevel` 내부 값을 꺼내와서 계산하고 있네요.
Enum의 내부 필드는 객체의 멤버변수와도 같습니다. 

AttendanceWarningLevel 에게 질문을 해보는건 어떨까요?

yebink replied:
`"경고"`
-> `AttendanceWarningLevel.WARNING` 과 사실상 의미하는 바가 같으므로, enum 의 내부 필드로 선언해도 될 것 같네요!

YehyeokBang replied:
더욱 직관적이고 이해하기 쉬운 방법이네요! 👍 
변경해보겠습니다!

yebink commented on src/main/java/attendance/model/AttendanceStartTime.java:
```diff
@@ -0,0 +1,35 @@
+
+public enum AttendanceStartTime {
+
+    MON(DayOfWeek.MONDAY, LocalTime.of(13, 0)),
```
`줄여 쓰지 않는다(축약 금지).` 요구사항을 지켜볼까요?

YehyeokBang replied:
처음에는 `DayOfWeek.MONDAY` 와 구분하기 위해 축약하여 사용했지만,

다시 보니 각 Enum이 자체적인 네임스페이스를 가지는 장점 을 활용하지 못하고 있었네요. (요구사항도 지키지 못하고)
축약 없이 명확하게 작성하겠습니다. 감사합니다!

yebink commented on src/main/java/attendance/model/Attendances.java:
```diff
@@ -0,0 +1,69 @@
+        return attendance;
+    }
+
+    public Optional<Attendance> findByCrewAndDate(Crew crew, LocalDate date) {
+        return attendances.stream()
+                .filter(attendance -> attendance.equals(new Attendance(crew, LocalDateTime.of(date, LocalTime.MIN))))
+                .findFirst();
+    }
```
Optional 은 꼭 필요한 상황이 아니면 그닥 추천하지 않습니다.
특히 Optional 을 리턴하게 되면, 해당 메서드를 사용하는 모든 곳에서 Optional 내부 값이 없는 경우에 대한 처리가 필요해지고, 그만큼 코드가 복잡해집니다.

Optional 을 사용하지 않도록 코드를 변경해볼까요~?

참고 : https://server-technology.tistory.com/487

YehyeokBang replied:
> Optional 은 꼭 필요한 상황이 아니면 그닥 추천하지 않습니다.
특히 Optional 을 리턴하게 되면, 해당 메서드를 사용하는 모든 곳에서 Optional 내부 값이 없는 경우에 대한 처리가 필요해지고, 그만큼 코드가 복잡해집니다.

맞아요. 처음 반환값을 Optional로 선언했을 때,
매개변수로 Optional을 받는 메서드까지 만들었던 기억이 있어요. 😢
Optional을 사용하면 "null은 피하겠다!"라는 생각으로 책임을 전가하는 느낌도 들었습니다.

그래서 제가 Optional을 사용하게 된 이유를 다시 고민해보았습니다.

## Optional을 사용하게 된 이유

요구 사항
- 전날까지의 모든 출석 기록을 출력해야 한다.
- 출석이 가능했던 날(주말·공휴일 제외)은 모두 포함해야 한다.
- 출석 가능한 날이지만 기록이 없는 경우, 결석으로 처리해 출력해야 한다.

이 요구 사항을 만족하기 위해 두 가지 방식을 생각했습니다. (2번 방법을 선택했습니다.)

1. 매일 결석 기록을 미리 생성한 후, 출석이 있으면 수정하기
    → 출석 기록이 없는 날도 미리 Attendance 객체로 관리

2. ✅ 실제 출석만 저장하고, 결석은 출력 시 연산하여 처리하기

이 과정에서, 특정 날짜의 출석 정보를 조회하는 메서드가 필요했습니다.

### (참고) 출석 객체로 결석을 표현하지 못한 이유

```java
public class Attendance {
    private final Crew crew;
    private final LocalDateTime dateTime;

    // 날짜와 관련된 연산을 수행하는 메서드들이 존재
}
```

- dateTime이 null이 되면, 출석 객체 내부에서 수행하는 모든 연산이 null-safe하게 작성되어야 합니다.

즉, 출석 객체를 "결석"을 나타내는 용도로도 사용하려면, 
내부적으로 계속해서 null 체크를 해야 하는 문제가 생길 수 있습니다.
그래서 "출석이 없다"는 상태를 Attendance 객체 내부에서 표현하지 않았습니다.

### 다시 돌아와서..

이때 `findByCrewAndDate(Crew crew, LocalDate date)` 메서드를 구현하게 되었고,
출석이 있는 경우 Attendance 객체를 반환하고, `없는 경우를 어떻게 처리할지 고민`하게 되었습니다.

### 출석 기록이 없는 경우는?

처음에는 `findByCrewAndDate()` 메서드를 다음과 같이 나누는 방법을 고민했습니다.

`existsAttendanceByCrewAndDate(Crew crew, LocalDate date)`
→ 출석 여부만 확인하는 메서드를 만들고 먼저 사용하도록 만들고
`findByCrewAndDate(Crew crew, LocalDate date)`
→ 존재하는 경우 Attendance를 반환, 없으면 예외를 발생시킨다.

그러나 이 방식은 존재 여부 판단 메서드를 먼저 호출하도록 강제할 수단이 없었습니다.
추가로 존재하지 않는 경우 예외가 발생한다는 것은  `출석 가능한 날에 기록이 없다면 결석이다`라는 요구사항에 맞지 않는 것 같았습니다.
  - 없더라도 예외가 아니라 사실은 결석이라는 무언가가 나와야 하지 않을까?라는 생각이었습니다.

그래서 Optional을 사용하면 `출석(기록)이 없을 수도 있다`는 개념을 명확히 전달할 수 있을 것 같다는 생각을 하게 되었습니다.

### 고민한 점
- Optional은 불필요하게 사용하면 코드 복잡성이 증가한다.
  - 맞습니다. 여러 관리 요소가 추가됩니다.
- null을 직접 반환하거나 예외를 던지는 방식이 더 나은가?
  - null을 직접 반환하는 것을 더 싫어하는 것 같습니다.
- exists~() 메서드를 먼저 호출하도록 강제하는 것이 직관적인가?
  - 강제할 수단이 떠오르지 않았습니다.

그래서 현재 구조에서 Optional을 없애려면 `출석이 없을 수도 있음`을 더 자연스럽게 표현하는 방법이 필요하다고 생각했습니다.
그렇지 않다면 Optional을 유지하는 것이 더 나은 선택일 수도 있다고 판단했던 것 같습니다.

---

아직 떠오르진 않았지만, 출석 객체에 대한 다른 표현 방법을 고민해보고 있습니다.

제 생각 흐름에 대한 조언을 부탁드려도 될까요? 더 나은 방식에 대한 조언도 좋습니다! 😄

YehyeokBang replied:
## 생각난 대안

기존에는 `LocalDateTime`을 사용하여 날짜와 시간을 함께 관리했습니다.
그렇기 때문에 해당 필드에서 null을  허용했을 때의 여파가 컸습니다.

그래서 `LocalDate(날짜)`와 `LocalTime(시간)`을 분리하여, 
null이 가능한 범위를 `시간`으로 좁혔습니다.

출석하지 않은 경우, 시간 필드만 null로 설정하여 
`출석 가능한 날인데 기록이 없어서 결석`을 더 명확하게 표현할 수 있게 되었습니다.
(기존에는 Optional로 결석 여부를 판단했지만, 이제는 시간 값이 없는 것으로 직접 표현 가능)

덕분에 Optional 반환을 제거할 수 있었고, 관련 코드도 훨씬 간결해졌습니다.
또한, 참고 자료를 확인하여 orElse() 대신 orElseGet()을 활용하여 성능 이점도 고려했습니다.

결과적으로, 기존보다 출석 데이터의 표현이 더 명확해지고, 코드도 단순해지는 효과를 얻을 수 있었습니다. 👍

yebink commented on src/test/java/attendance/model/AttendanceStartTimeTest.java:
```diff
@@ -0,0 +1,41 @@
+            "FRIDAY, 10:00",
+    })
+    void findDayOfWeekTest(DayOfWeek dayOfWeek, LocalTime startTime) {
+        LocalTime attendanceStartTime = AttendanceStartTime.findDayOfWeek(dayOfWeek);
+        assertThat(attendanceStartTime)
+                .isEqualTo(startTime);
```
테스트에  `// given`, `// when`, `// then` 주석을 달면 가독성을 높일 수 있어요 🙂

yebink reviewed (changes requested):
안녕하세요 벨로~

출석 미션 구현하시느라 고생 많으셨어요 👍 
주신 질문에 대한 답변과 코멘트 몇개 남겼으니 확인하시고 다시 요청주세요~

yebink commented:
안녕하세요 벨로!
출석 미션 리뷰를 맡은 엘리입니다.
만나서 반갑습니다 😄 

주신 질문에 대한 답변 드려요.

* **View에서 Enum 참조 관련**
  * 우선 Enum 에 로직을 아주 잘 녹여내셨네요! 💯
  * 가독성도 좋고, 메인 로직이 한곳에 정의되어 있어서 코드를 이해하기 편했습니다. 👍
  * 지금의 요구사항에서는 DTO 를 사용했을 때 이점이 없다는 부분에 동의합니다.
  * 관련해서 [리뷰](https://github.com/woowacourse/java-attendance/pull/45#discussion_r1963671447) 드렸으니 참고해보세요.
* **null 필드 관련**
  * 좋은 고민이네요 👏 
  * null 을 피할 수 있는 방법..은 아쉽지만 자바 진영에서는 아직 어려운 이야기입니다ㅠ 저도 벨로가 구현한 방식처럼 최대한 null 을 만들지 않는 것을 선호합니다. `출석 객체를 미리 생성하는 방법` 의 또다른 문제는 서비스를 확장할 때 발생합니다. 프로그램 규모가 커지면 어떨까요? 사용자가 100만 명이 되면, 또는 서비스를 5년, 10년 이상 하게 되면 그 많은 null 데이터를 저장하는 것은 큰 낭비이겠죠?
  * 결국 null 사용을 지양하되, 발생할 수 있는 부분은 꼼꼼히 방어하는 것이 최선인 것 같습니다 😄 
* **출석을 수정한다 vs 새로운 출석을 만들어 대체한다**
  * 정답은 없습니다. 이건 케바케가 정말 심하고 현업에서도 여전히 고민하는 문제에요 ㅎㅎ 
  * 불변 객체를 유지하려고 하신 건 (수정) 이 아닌 (삭제 후 생성) 에 대한 충분한 근거가 된다고 생각해요.
  * 그러나 attendances 가 set 이 아닌 데이터베이스 로 관리되어야 한다면, 또 다른 문제에요.
    * `수정 시 쿼리 1번 발생` vs `삭제 후 생성 시 쿼리 2번 발생`
    * db에 쿼리를 날리는 비용과 불변 객체를 유지하는 비용 을 두고 고민이 필요합니다.
  * 우선 지금의 요구사항에서는 충분히 좋은 선택이었습니다!
  * 무엇보다도 이런걸 고민한다는 것은 정말 좋은 자세에요 👏 💯

yebink commented:
실패하는 테스트도 확인 부탁드립니다~ 😄

YehyeokBang committed:
fix: 불필요한 괄호 제거

YehyeokBang committed:
fix: 테스트 코드 수정

- 기존 오늘까지 계산하던 코드를 전날까지로 바꾸면서 테스트가 실패했습니다.
- 기대하는 값을 전날까지의 결과로 변경합니다.

YehyeokBang committed:
refactor: Enum 인스턴스에게 요청하는 방식으로 변경

- 기존 각 Enum이 의미하는 바를 출력하기 위해 인스턴스를 비교하고 출력하던 방식에서
  Enum 필드에 의미를 추가하고 직접 반환하도록 수정합니다.

YehyeokBang committed:
refactor: 축약했던 이름을 수정합니다.

- 축약하지 않는다는 요구사항을 지키도록 수정합니다.

YehyeokBang committed:
refactor: 프로그램 실행 시간 주입으로 변경 및 리팩토링

- 기존 컨트롤러 내부에서 현재 시간을 가져오던 방식에서 컨트롤러 외부에서 주입하도록 변경합니다.
- now라고 표기했던 기준 시간이라고 명확하게 표현합니다.
- 메서드명 및 변수명을 더욱 명확하게 변경합니다.
- 시간 포맷 기능을 유틸 클래스로 분리합니다.
- 내부 클래스로 정의되어 있던 dto를 분리합니다.

YehyeokBang committed:
refactor: 크루의 닉네임 포장

- 크루의 닉네임을 포장하며 스스로 검증하도록 합니다.
- 기타 매직 넘버를 상수 처리합니다.
- 테스트 코드에서 LocalDate.now를 사용하며 주말에는 테스트가 깨지는 문제를 수정합니다.

YehyeokBang committed:
refactor: 테스트 코드 리팩토링

- 자주 사용되는 객체를 위해 TestFixture를 추가합니다.
- given, when, then 구조로 정렬합니다.

YehyeokBang committed:
refactor: 출석 객체 날짜와 시간 분리

- 기존 LocalDateTime 객체를 두개로 분리합니다.
- null 관리 대상 범위를 줄이면서 Optional을 사용하지 않도록 변경합니다.
- 출석 객체를 확인할 때 null이 아니라 기록되지 않았는지 확인하도록 합니다.

YehyeokBang commented:
안녕하세요 엘리!

꼼꼼한 리뷰 덕분에 다시 돌아보고 학습할 수 있던 좋은 기회가 되었습니다. 🙇

## 출석을 수정한다 vs 새로운 출석을 만들어 대체한다.

> 출석을 수정한다 vs 새로운 출석을 만들어 대체한다
`attendances` 가 `set 이 아닌 데이터베이스`로 관리되어야 할 때 발생할 수 있는 문제,
수정 시 쿼리 1번 발생 vs 삭제 후 생성 시 쿼리 2번 발생
db에 쿼리를 날리는 비용과 불변 객체를 유지하는 비용 을 두고 고민이 필요합니다.

엘리의 경험을 바탕으로 생생한 답변을 받을 수 있어서 좋았습니다. 👍 👍 
언제나 정답은 없고, 현재 상황과 선택 근거가 중요하다는 것을 다시 깨닫게 되는 좋은 답변이었어요. 😄 

## 피드백 반영

### Optional의 단점 (많은 시행 착오...)

> Optional 은 꼭 필요한 상황이 아니면 그닥 추천하지 않습니다.
특히 Optional 을 리턴하게 되면, 해당 메서드를 사용하는 모든 곳에서 Optional 내부 값이 없는 경우에 대한 처리가 필요해지고, 그만큼 코드가 복잡해집니다.

현재 구조에서 Optional을 제거하려고 여러 방법을 고민했지만, 
결국 대체 방법들도 비슷한 단점을 가지고 있다고 느꼈습니다.

Optional을 없애려면 `출석이 없을 수도 있음`을 더 자연스럽게 표현하는 방법이 필요합니다.
하지만 그런 방법이 없다면, 오히려 Optional을 유지하는 것이 더 나은 선택일 수도 있다고 생각했습니다.
(이 구조에서 출석 객체는 '결석'을 표현할 수 없었기 때문에, Optional을 통해 출석 기록이 없다는 사실을 전달할 수 있었습니다.)

[리뷰 답변 본문](https://github.com/woowacourse/java-attendance/pull/45#discussion_r1966669816)

### 단점을 인지하고도 변경하지 않는 것은 모순이다

그렇지만, Optional의 단점을 알고도 그대로 두는 것은 모순이라고 생각했습니다.
그래서 리뷰에서 제공된 참고 자료를 읽고, 
답변을 정리하는 과정에서 떠오른 방법으로 구현을 변경해 보았습니다.

✅ 변경 내용

기존에는 `LocalDateTime`을 사용하여 날짜와 시간을 함께 관리했는데, 
이 부분 때문에 null을 허용했을 때의 여파가 컸습니다.
그래서 `LocalDate(날짜)`와 `LocalTime(시간)`을 분리하여, null이 가능한 범위를 시간으로 좁혔습니다.

출석하지 않은 경우, 시간 필드만 null로 설정하여 `출석 가능한 날인데 기록이 없어서 결석`을 더 명확하게 표현할 수 있게 되었습니다.
(기존에는 Optional로 결석 여부를 판단했지만, 이제는 시간 값이 없는 것으로 직접 표현 가능)
- 오히려 없는 것을 없다고 표현하는 것이 명확할 수 있겠다는 생각을 했습니다.
  - 만약 null을 막기 위해 `LocalTime NO_RECORDED_TIME = LocalDate.MIN;`처럼 
  기록된 시간이 없다는 것을 상수로 선언하고 활용하는 것은 그 순간 코드를 읽는 것은 쉽겠지만,
   확장할 때 많은 문제가 발생할 수 있다고 생각했습니다. (작성자만 아는 콘텍스트가 생긴다는 점)

덕분에 Optional 반환을 제거할 수 있었고, 관련 코드도 훨씬 간결해졌습니다.
또한, 참고 자료를 확인하여 `orElse()` 대신 `orElseGet()`을 활용하여 성능 이점도 고려했습니다.

결과적으로, 기존보다 출석 데이터의 표현이 더 명확해지고, 코드도 단순해지는 효과를 얻을 수 있었습니다. 👍 

### 테스트 코드 // given, // when, // then 주석

- 테스트도 결국 코드이기에 읽고 이해하기 쉬워야 한다는 관점에서 공감하게 되어 반영하게 되었습니다.
- 추가적으로 반영하는 과정에서 자주 생성되는 객체나 중복되는 부분을 줄이고 가독성을 높이기 위해 
(상태를 공유하지 않는 객체 위주로) 학습 자료 힌트에 있던 TestFixture 방식을 활용했습니다.

### 모든 원시 값과 문자열을 포장한다.

- 크루의 닉네임을 객체로 포장하여, 닉네임 스스로 검증할 수 있도록 구현했습니다.
- 예외 메시지도 포장해야 하는가? 라는 고민이 들었습니다.
  - 예외 메시지는 대부분 한 번만 사용되며, 이를 별도로 포장했을 때의 장점을 크게 느끼지 못했습니다.
  - 현재로서는 예외 메시지를 포장하는 것이 불확실한 미래에 대한 대비일 뿐, 지금 코드를 더 명확하게 만들지는 않는다고 판단했습니다.
  - 물론, 예외 메시지를 일관되게 관리해야 할 상황이 온다면 도움이 될 수도 있습니다.
  - 하지만 그 필요성이 불확실한 상태에서 미리 적용하는 것은, 오히려 불필요한 코드가 될 위험이 있다는 점이 걱정되었습니다.
  
즉, 현재 상황에서 필요한 이유가 분명하지 않다면(느끼지 못한다면), 
무조건적인 포장은 오히려 복잡도를 높일 수 있다고 생각했어요. 
이에 관한 엘리의 생각이 궁금해요 😄  

사실 `요구사항`이기 때문에 높은 우선순위로 지키는 게 좋다고 생각합니다. 
그렇지만... 고민을 나누고 더 나은 방식을 찾아볼 수 있지 않을까? 라는 마음으로 질문하게 되었습니다.

### 파라미터명, 변수명, 메서드명 등

- 파라미터 now 대신, 기준 시간이라는 내용을 명확히 표현합니다.
- 더욱 명확한 표현과 `일관성 유지`를 위해 네이밍 패턴을 도입했습니다.
  - displayXXX() → 뷰(View) 객체를 사용하여 화면에 출력하는 메서드
  - view.printXXX() → 현재 요구사항인 콘솔 출력을 담당하는 메서드
  - readXXX() → 입력을 읽는 역할의 메서드
  - view.inputXXX() → 뷰 자체에서 입력을 처리하는 메서드
- 기존의 `logic()`처럼 추상적인 메서드명을 지양하고,
`processUserCommand()`, `executeCommand()` 등
 사용자의 명령을 수행하는 역할이 드러나도록 개선했습니다.

이러한 방식으로 네이밍을 정리하여, 
메서드의 역할을 직관적으로 이해할 수 있도록 개선해봤습니다.

감사합니다! 👍

YehyeokBang commented:
> 실패하는 테스트도 확인 부탁드립니다~ 😄

빠트린 요구사항을 마감 직전 급하게 수정하느라 테스트 코드를 확인하지 못했습니다. 😢 
이번에는 꼭 확인하고 요청드리겠습니다! 

<img width="571" alt="image" src="https://github.com/user-attachments/assets/60f590ce-7022-45e6-9de8-8648cc2f52e8" />

YehyeokBang commented on src/main/java/attendance/AttendanceProgram.java:
```diff
@@ -0,0 +1,21 @@
+        return new AttendanceController(new InputView(), new OutputView());
+    }
+
+    private static LocalDateTime initializeBaseTime() {
+        return LocalDateTime.now();
+    }
```
```suggestion
    private static LocalDateTime initializeBaseTime() {
        return LocalDateTime.of(2024,12,13, 10,1);
    }
```

- 특정 날짜가 아닌 새롭게 지정한 날짜에도 동작할 수 있도록 분리했었습니다.
  - 출력 예시처럼 하려면, 아래의 날짜를 12월 13일로 바꿔주시면 됩니다.

yebink commented on src/main/java/attendance/model/AttendanceType.java:
```diff
@@ -0,0 +1,39 @@
+        return Arrays.stream(values())
+                .filter(attendanceType -> attendanceType.isMatch.apply(calculateDifMinutes(startTime, attendanceTime)))
+                .findFirst()
+                .orElseThrow(() -> new IllegalArgumentException("판단할 수 없습니다."));
```
좀 더 구체적인 예외 메세지를 작성해볼까요?

yebink commented on src/test/java/attendance/model/AttendanceTestFixtures.java:
```diff
@@ -0,0 +1,33 @@
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class AttendanceTestFixtures {
```
👍

yebink commented on src/main/java/attendance/util/DateTimeUtil.java:
```diff
@@ -0,0 +1,41 @@
+    private static final DateTimeFormatter DATE_TIME_FORMATTER = ofPattern("MM월 dd일 E요일 HH:mm", KOREAN);
+    private static final DateTimeFormatter DATE_FORMATTER = ofPattern("MM월 dd일 E요일", KOREAN);
+    private static final DateTimeFormatter TIME_FORMATTER = ofPattern("HH:mm", KOREAN);
+    public static final String TIME_NOT_RECORDED = "--:--";
```
해당 변수는 OutputView에서만 사용하고 있습니다. 
적절한 위치를 고민해볼까요?

yebink commented:
> 즉, 현재 상황에서 필요한 이유가 분명하지 않다면(느끼지 못한다면), 무조건적인 포장은 오히려 복잡도를 높일 수 있다고 생각했어요. 이에 관한 엘리의 생각이 궁금해요 😄

동의합니다. 원시값을 포장하는게 때로는 가독성을 저하시킬 수 있다고 생각해요.
다만 학습의 측면에서 최대한 구현해보는 연습을 하는 것도 좋을 것 같습니다 🙂 

> Optional 관련

요구사항을 가장 효율적으로 구현할 방법에 대해 깊이 고민한 것이 느껴지네요 💯 
현재 요구사항에서 null 보다 Optional 을 사용했을 때 이점이 있다는 부분에 동의합니다!
그리고 기존 Optional을 리턴하던 코드에서, Optional 을 메서드 내부에서만 사용하는 방식으로 변경하셨기 때문에, 제가 말씀드린 Optional의 단점이 사라졌습니다.
잘 구현하셨어요 👍

yebink reviewed (approved):
안녕하세요 벨로~

피드백 잘 반영해주셨네요 💯 👍 
추가 피드백 조금 드렸으니 step2 에서 같이 반영해주세요!
머지하겠습니다 🙂
